presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ApplyDynamicFilters.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.SystemSessionProperties;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.JoinNode.EquiJoinClause;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.DeferredSymbolReference;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.ExpressionUtils.combineConjuncts;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.EQUAL;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class ApplyDynamicFilters
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Context context)
+    {
+        if (!SystemSessionProperties.isDynamicPartitionPruningEnabled(context.getSession())) {
+            return Optional.empty();
+        }
+
+        if (!(node instanceof JoinNode)) {
+            return Optional.empty();
+        }
+
+        JoinNode join = (JoinNode) node;
+
+        if (join.getType() != INNER) {
+            return Optional.empty();
+        }
+
+        List<EquiJoinClause> criteria = join.getCriteria();
+        Assignments existingAssignments = join.getDynamicFilterAssignments();
+
+        List<EquiJoinClause> unprocessedClauses = criteria.stream()
+                .filter(c -> !existingAssignments.getExpressions().contains(c.getRight().toSymbolReference()))
+                .collect(toImmutableList());
+
+        if (unprocessedClauses.isEmpty()) {
```

Comment:
What about case when `existingAssignments` contains entries for `expressions` which are no longer parts of `equiJoinClauses`.
Is that possible? We should probably disallow or filter out such assignments in `JoinNode` constructor.